File: app_manager/appmanager/src/main/java/app/editors/manager/ui/fragments/login/PersonalSignUpFragment.java
package app.editors.manager.ui.fragments.login;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.appcompat.widget.AppCompatButton;
import androidx.appcompat.widget.AppCompatEditText;

import moxy.presenter.InjectPresenter;
import com.google.android.material.textfield.TextInputLayout;

import app.editors.manager.R;
import app.editors.manager.mvp.presenters.login.PersonalSignUpPresenter;
import app.editors.manager.mvp.views.login.PersonalRegisterView;
import app.editors.manager.ui.fragments.base.BaseAppFragment;
import app.editors.manager.ui.views.edits.BaseWatcher;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.OnEditorAction;
import butterknife.Unbinder;
import lib.toolkit.base.managers.utils.StringUtils;
import lib.toolkit.base.ui.dialogs.common.CommonDialog;

public class PersonalSignUpFragment extends BaseAppFragment implements PersonalRegisterView {

    public static final String TAG = PersonalSignUpFragment.class.getSimpleName();

    private static final String TAG_DIALOG_WAITING = "TAG_DIALOG_WAITING";
    private static final String TAG_DIALOG_INFO = "TAG_DIALOG_INFO";

    protected Unbinder mUnbinder;
    @BindView(R.id.login_personal_portal_email_edit)
    protected AppCompatEditText mLoginPersonalPortalEmailEdit;
    @BindView(R.id.login_personal_portal_email_layout)
    protected TextInputLayout mLoginPersonalPortalEmailLayout;
    @BindView(R.id.login_personal_signup_button)
    protected AppCompatButton mLoginPersonalSignupButton;

    @InjectPresenter
    PersonalSignUpPresenter mPersonalSignUpPresenter;

    private FieldsWatcher mFieldsWatcher;

    public static PersonalSignUpFragment newInstance() {
        return new PersonalSignUpFragment();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        final View view = inflater.inflate(R.layout.fragment_login_personal_signup, container, false);
        mUnbinder = ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        init(savedInstanceState);
    }

    @Override
    public void onAcceptClick(CommonDialog.Dialogs dialogs, @Nullable String value, @Nullable String tag) {
        super.onAcceptClick(dialogs, value, tag);
        if (tag != null) {
            switch (tag) {
                case TAG_DIALOG_INFO:
                    getActivity().finish();
                    break;
            }
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mUnbinder.unbind();
    }

    @Override
    public void onCancelClick(CommonDialog.Dialogs dialogs, @Nullable String tag) {
        super.onCancelClick(dialogs, tag);
        if (tag != null) {
            switch (tag) {
                case TAG_DIALOG_WAITING:
                    mPersonalSignUpPresenter.cancelRequest();
                    break;
                case TAG_DIALOG_INFO:
                    hideDialog();
                    break;
            }
        }
    }

    @OnClick(R.id.login_personal_signup_button)
    public void onSignUpClick() {
        hideKeyboard(mLoginPersonalPortalEmailEdit);
        final String email = mLoginPersonalPortalEmailEdit.getText().toString();
        if (StringUtils.isEmailValid(email)) {
            showWaitingDialog(getString(R.string.dialogs_sign_in_register_portal), getString(R.string.dialogs_common_cancel_button), TAG_DIALOG_WAITING);
            mPersonalSignUpPresenter.registerPortal(email);
        } else {
            setMessage(R.string.errors_email_syntax_error, true);
        }
    }

    @OnEditorAction(R.id.login_personal_portal_email_edit)
    protected boolean actionKeyPress(final TextView v, final int actionId, final KeyEvent event) {
        if (actionId == EditorInfo.IME_ACTION_DONE) {
            onSignUpClick();
            return true;
        }

        return false;
    }

    @Override
    public void onError(@Nullable String message) {
        hideDialog();
        setMessage(message, true);
    }

    @SuppressLint("StringFormatInvalid")
    @Override
    public void onRegisterPortal() {
        showQuestionDialog(getString(R.string.dialogs_question_personal_confirm_link_title),
                getString(R.string.login_personal_signup_dialog_info, mLoginPersonalPortalEmailEdit.getText().toString()),
                getString(R.string.dialogs_question_personal_confirm_accept), getString(R.string.dialogs_question_personal_confirm_cancel),
                TAG_DIALOG_INFO);
    }

    private void init(final Bundle savedInstanceState) {
        setActionBarTitle(getString(R.string.login_personal_signup_title));
        showKeyboard(mLoginPersonalPortalEmailEdit);
        mFieldsWatcher = new FieldsWatcher();
        mLoginPersonalPortalEmailEdit.addTextChangedListener(mFieldsWatcher);
        mLoginPersonalSignupButton.setEnabled(false);
        setMessage(R.string.login_personal_signup_edit_info, false);
    }

    private void setMessage(final String message, final boolean isError) {
        mLoginPersonalPortalEmailLayout.setErrorTextAppearance(isError ? R.style.TextInputErrorRed : R.style.TextInputErrorGrey);
        mLoginPersonalPortalEmailLayout.setError(message);
    }

    private void setMessage(@StringRes final int resId, final boolean isError) {
        setMessage(getString(resId), isError);
    }

    private class FieldsWatcher extends BaseWatcher {

        @Override
        public void onTextChanged(CharSequence text, int start, int before, int count) {
            mLoginPersonalPortalEmailLayout.setErrorTextAppearance(R.style.TextInputErrorGrey);
            mLoginPersonalPortalEmailLayout.setError(getString(R.string.login_personal_signup_edit_info));
            final String email = mLoginPersonalPortalEmailEdit.getText().toString();
            if (!"".equals(email)) {
                mLoginPersonalSignupButton.setEnabled(true);
            } else {
                mLoginPersonalSignupButton.setEnabled(false);
            }
        }
    }

}

File: app_manager/appmanager/src/main/java/app/editors/manager/ui/fragments/login/PersonalSignUpFragment.kt
package app.editors.manager.ui.fragments.login

import android.annotation.SuppressLint
import android.content.Context
import android.os.Bundle
import android.text.Spanned
import android.text.method.LinkMovementMethod
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import androidx.annotation.StringRes
import androidx.fragment.app.viewModels
import app.editors.manager.R
import app.editors.manager.app.appComponent
import app.editors.manager.databinding.FragmentLoginPersonalSignupBinding
import app.editors.manager.mvp.presenters.login.PersonalSignUpPresenter
import app.editors.manager.mvp.views.login.PersonalRegisterView
import app.editors.manager.ui.fragments.base.BaseAppFragment
import app.editors.manager.ui.views.edits.BaseWatcher
import app.editors.manager.viewModels.login.RemoteUrlViewModel
import lib.toolkit.base.ui.dialogs.common.CommonDialog.Dialogs
import moxy.presenter.InjectPresenter

class PersonalSignUpFragment : BaseAppFragment(), PersonalRegisterView {

    companion object {
        val TAG: String = PersonalSignUpFragment::class.java.simpleName

        private const val TAG_DIALOG_WAITING = "TAG_DIALOG_WAITING"
        private const val TAG_DIALOG_INFO = "TAG_DIALOG_INFO"

        fun newInstance(): PersonalSignUpFragment {
            return PersonalSignUpFragment()
        }
    }

    @InjectPresenter
    lateinit var personalSignUpPresenter: PersonalSignUpPresenter

    private val urlsViewModel: RemoteUrlViewModel by viewModels()

    private var fieldsWatcher: FieldsWatcher? = null

    private var viewBinding: FragmentLoginPersonalSignupBinding? = null

    override fun onAttach(context: Context) {
        super.onAttach(context)
        context.appComponent.inject(urlsViewModel)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        viewBinding = FragmentLoginPersonalSignupBinding.inflate(inflater)
        return viewBinding?.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        init()
    }

    override fun onAcceptClick(dialogs: Dialogs?, value: String?, tag: String?) {
        super.onAcceptClick(dialogs, value, tag)
        if (tag != null) {
            when (tag) {
                TAG_DIALOG_INFO -> requireActivity().finish()
            }
        }
    }

    override fun onCancelClick(dialogs: Dialogs?, tag: String?) {
        super.onCancelClick(dialogs, tag)
        if (tag != null) {
            when (tag) {
                TAG_DIALOG_WAITING -> personalSignUpPresenter.cancelRequest()
                TAG_DIALOG_INFO -> hideDialog()
            }
        }
    }

    override fun onWaitingDialog() {
        showWaitingDialog(
            getString(R.string.dialogs_sign_in_register_portal),
            getString(R.string.dialogs_common_cancel_button),
            TAG_DIALOG_WAITING
        )
    }

    override fun onMessage(message: Int) {
        setMessage(message = getString(message), true)
    }

    private fun onSignUpClick() {
        hideKeyboard(viewBinding?.loginPersonalPortalEmailEdit)
        val email = viewBinding?.loginPersonalPortalEmailEdit?.text.toString()
        personalSignUpPresenter.checkMail(email = email)
    }

    private fun actionKeyPress(actionId: Int): Boolean {
        if (actionId == EditorInfo.IME_ACTION_DONE) {
            onSignUpClick()
            return true
        }
        return false
    }

    override fun onError(message: String?) {
        hideDialog()
        setMessage(message, true)
    }

    @SuppressLint("StringFormatInvalid")
    override fun onRegisterPortal() {
        showQuestionDialog(
            getString(R.string.dialogs_question_personal_confirm_link_title),
            getString(
                R.string.login_personal_signup_dialog_info,
                viewBinding?.loginPersonalPortalEmailEdit?.text.toString().replace(PersonalSignUpPresenter.TAG_INFO, "")
            ),
            getString(R.string.dialogs_question_personal_confirm_accept),
            getString(R.string.dialogs_question_personal_confirm_cancel),
            TAG_DIALOG_INFO
        )
    }

    private fun init() {
        fieldsWatcher = FieldsWatcher()
        initListeners()
        setActionBarTitle(getString(R.string.login_personal_signup_title))
        showKeyboard(viewBinding?.loginPersonalPortalEmailEdit)
        viewBinding?.loginPersonalSignupButton?.isEnabled = false
        setMessage(R.string.login_personal_signup_edit_info, false)
        urlsViewModel.remoteUrls.observe(viewLifecycleOwner) { text: Spanned? ->
            text?.let {
                viewBinding?.terms?.termsTextView?.movementMethod = LinkMovementMethod.getInstance()
                viewBinding?.terms?.termsTextView?.text = text
            }
        }
    }

    private fun initListeners() {
        viewBinding?.loginPersonalSignupButton?.setOnClickListener {
            onSignUpClick()
        }

        viewBinding?.loginPersonalPortalEmailEdit?.setOnEditorActionListener { _, actionId, _ ->
            actionKeyPress(actionId)
        }

        viewBinding?.terms?.termsCheckbox?.setOnCheckedChangeListener { _, isChecked ->
            viewBinding?.loginPersonalSignupButton?.isEnabled = isChecked && viewBinding?.loginPersonalPortalEmailEdit?.text?.isNotEmpty() == true
        }

        viewBinding?.loginPersonalPortalEmailEdit?.addTextChangedListener(fieldsWatcher)
    }

    private fun setMessage(message: String?, isError: Boolean) {
        viewBinding?.loginPersonalPortalEmailLayout?.apply {
            setErrorTextAppearance(if (isError) lib.toolkit.base.R.style.TextInputErrorRed else lib.toolkit.base.R.style.TextInputErrorGrey)
            error = message
        }
    }

    private fun setMessage(@StringRes resId: Int, isError: Boolean) {
        setMessage(getString(resId), isError)
    }

    private inner class FieldsWatcher : BaseWatcher() {
        override fun onTextChanged(text: CharSequence, start: Int, before: Int, count: Int) {
            setMessage(R.string.login_personal_signup_edit_info, false)
            val email = viewBinding?.loginPersonalPortalEmailEdit?.text.toString()
            viewBinding?.loginPersonalSignupButton?.isEnabled = "" != email && viewBinding?.terms?.termsCheckbox?.isChecked == true
        }
    }
}